File: JavaScript/ExampleTests/common/OrderItemTyped.js
import { OrderItemString } from "./OrderItemString.js";
import * as _json from "./json.js";

export class OrderItemTyped {
  constructor(name = "", quantity = 0, price = "", itemTotal = "") {
    this.name = name;
    this.quantity = quantity;
    this.price = price;
    this.itemTotal = itemTotal;
  }

  static fromStringObj(s) {
    return new OrderItemTyped(
      s.name,
      s.quantity !== "" ? Number(s.quantity) : 0,
      s.price,
      s.itemTotal
    );
  }

  toJsonValue() {
    return {
      name: this.name == null ? null : String(this.name),
      quantity: this.quantity,
      price: this.price == null ? null : String(this.price),
      itemTotal: this.itemTotal == null ? null : String(this.itemTotal),
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new OrderItemTyped(
      new SimpleText(_json.asString(_json.require(m, "name"), "name")),
      _json.asInt(_json.require(m, "quantity"), "quantity"),
      new Dollar(_json.asString(_json.require(m, "price"), "price")),
      new Dollar(_json.asString(_json.require(m, "itemTotal"), "itemTotal"))
    );
  }

  static fromJSON(text) { return OrderItemTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "OrderItemTyped");
    return raw.map((e) => OrderItemTyped.fromJsonValue(e));
  }

  toString() {
    return `Name=${this.name}, Quantity=${this.quantity}, Price=${this.price}, ItemTotal=${this.itemTotal}`;
  }

  equals(other) {
    if (!(other instanceof OrderItemTyped)) return false;
    return this.name === other.name
      && this.quantity === other.quantity
      && this.price === other.price
      && this.itemTotal === other.itemTotal;
  }
}

File: Python/ExampleTests/common/order_item_typed.py
from . import json_util as _json
from .order_item_string import OrderItemString

class OrderItemTyped:
    def __init__(self, name: str = '', quantity: int = 0, price: str = '', item_total: str = ''):
        self.name = name
        self.quantity = quantity
        self.price = price
        self.item_total = item_total

    @classmethod
    def from_string_obj(cls, s: OrderItemString) -> 'OrderItemTyped':
        return cls(
            s.name,
            int(s.quantity) if s.quantity else 0,
            s.price,
            s.item_total
        )

    def to_json_value(self) -> dict:
        return {
            'name': self.name,
            'quantity': self.quantity,
            'price': self.price,
            'item_total': self.item_total,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'OrderItemTyped':
        return cls(
            _json.as_str(_json.require(m, 'name'), 'name'),
            _json.as_int(_json.require(m, 'quantity'), 'quantity'),
            _json.as_str(_json.require(m, 'price'), 'price'),
            _json.as_str(_json.require(m, 'item_total'), 'item_total')
        )

    @classmethod
    def from_json(cls, text: str) -> 'OrderItemTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'OrderItemTyped')
        return [cls.from_json_value(e) for e in raw]
